### ISSUE TITLE
Foreign key randomly filled

### ISSUE THREAD

**Issue by AhmedNetDev** [borisdj/EFCore.BulkExtensions#547]
Hi, i m using the library to bulk graphs to database and i m facing issues with Foreign key not always filled  when i launch the bulk with a list but whith only one graph passed to the bulk  the one-to-one relation is correctly filled.

Did you have such issue or do you want that i send you my model ?

**Comment by borisdj** [comment]
When doing Bulk to multiple tables that are related you can either call Bulk ops separately for each where for parent configure **SetOutputIdentity** while for child set FKs manually before sending list to Bulk method.
Another way would be to configure **IncludeGraph** and call Bulk just once for parent, which will automatically do the rest.
Both options are explained in the ReadMe with example of parent-child FK related tables.

**Comment by AhmedNetDev** [comment]
Hi @borisdj, i m already using the call with includeGraph like this :  await _context.BulkInsertOrUpdateAsync(tranches.ToList(), new BulkConfig
                {
                    IncludeGraph = true,
                });

below example of model 👍 
 public Class RootEntity
{
         public int Id { get; set; }
        public OwnedType Owned{ get; set; }
}
 public Class OwnedType 
{
     int? ChildId { get; set; }
     public ChildEntity Child{ get; set; }
     public string Field1 { get; set; }
}

and below the configuration 
        public void Configure(EntityTypeBuilder<RootEntity> builder)
        {
            builder.Property(d => d.Id).ValueGeneratedNever();

            builder
              .OwnsOne(t => t.Owned, bd =>
              {
                  bd.HasOne(d => d.Child).WithMany().HasForeignKey(d => d.ChildId ).IsRequired(false);
                  bd.Property(d => d.Field1 ).HasMaxLength(50);
              });
       }

I m using OwnsOne without separate table so i don't need additional call to bulk. 

Which is specific here is that the owned type has a HasOne relation and my question was about the ChieldId foreign key which is not always filled correctly. 

Is the library support HasOne inside a OwnsOne configuration ? Is there any problem with my configuration ?

I m using the last version of the library (5.1.7) in a .Net5 project

Thank you in adanvance for your help

**Comment by AhmedNetDev** [comment]
To complete the example above : 

    public class ChildEntity 
    {
        public int Id { get; set; }
        public string Name { get; set; }
    }

    public class ChildEntityConfiguration : IEntityTypeConfiguration<ChildEntity>
    {
        public void Configure(EntityTypeBuilder<ChildEntity> builder)
        {
            builder.Property(d => d.Id).ValueGeneratedNever();
            builder.Property(d => d.Name);
        }
    }

**Comment by borisdj** [comment]
That strcuture is not suported currently.
Try instead separate calls:
```C#
var tranches = new List<RootEntity>
{
    new RootEntity {
        Id = 1,
        Owned = new OwnedType
        {
            Field1 = "F1",
            ChildId = 1,
            Child = new ChildEntity { Id = 1, Name = "F1C1" }
        }
    },
    new RootEntity {
        Id = 2,
        Owned = new OwnedType
        {
            Field1 = "F2",
            ChildId = 2,
            Child = new ChildEntity { Id = 2, Name = "F2C2" }
        }
    }
};

using var context = new TestContext(ContextUtil.GetOptions());

await context.BulkInsertOrUpdateAsync(tranches.Select(a => a.Owned.Child).ToList());
await context.BulkInsertOrUpdateAsync(tranches);
```

**Comment by AhmedNetDev** [comment]
Unfortunaltly, it's not working.  The model that i mentioned is quite simple, in my real model i have many other relationship in RootEntity with hasmany etc.   That's why i found the includGraph very userful.

What i can't explain is that all is working fine if  bulk only one graph each time. For example, the bad test code below works 👍 

foreach(var rootEntity in rootEntities)
{
      await context.BulkInsertOrUpdateAsync(new List<RootEntity>() {rootEntity}, new BulkConfig
      {
           IncludeGraph = true,
      }););
}

but when i bulk athe home list, the ChildId foreign key will be set to null (that's why i mentioned in the title of the issue the "randomly" word). 

I hope i was clear. I tried also to tranform the OwnsOne to HaseOne to test but the result is the same.   not simple to reproduce like behavior.

**Comment by borisdj** [comment]
IncludeGraph was PR, @maitlandmarshall again for you if find the time.

**Comment by maitlandmarshall** [comment]
@AhmedNetDev I am unable to replicate the exact issue you are describing. Are you sure an exception is not being thrown?

When I use OwnsOne it completely fails with an exception.
`'The MERGE statement conflicted with the FOREIGN KEY constraint "FK_RootEntities_ChildEntity_Owned_ChildId". The conflict occurred in database "EFCoreBulkTest_Issue547", table "dbo.ChildEntity", column 'Id'.`

This is because the OwnedType confuses the topological sort and inserts the entities in the wrong order.

Everything imports successfully when I use HasOne instead of OwnsOne.
```cs
namespace EFCore.BulkExtensions.Tests.IncludeGraph
{
    public class RootEntity
    {
        public int Id { get; set; }
        public OwnedType Owned { get; set; }
    }

    public class OwnedType
    {
        public int Id { get; set; }
        public int? ChildId { get; set; }
        public ChildEntity Child { get; set; }
        public string Field1 { get; set; }
    }

    public class ChildEntity
    {
        public int Id { get; set; }
        public string Name { get; set; }
    }

    public class Issue547DbContext : DbContext
    {
        public Issue547DbContext([NotNullAttribute] DbContextOptions options) : base(options)
        {
        }

        public DbSet<RootEntity> RootEntities { get; set; }

        protected override void OnModelCreating(ModelBuilder modelBuilder)
        {
            base.OnModelCreating(modelBuilder);

            modelBuilder.Entity<RootEntity>(cfg =>
            {
                cfg.Property(y => y.Id).ValueGeneratedNever();
                cfg.HasOne(y => y.Owned);
            });

            modelBuilder.Entity<OwnedType>(own =>
            {
                own.Property(y => y.Id).ValueGeneratedNever();
                own.HasOne(y => y.Child).WithMany().HasForeignKey(y => y.ChildId).IsRequired(false);
                own.Property(y => y.Field1).HasMaxLength(50);
            });

            modelBuilder.Entity<ChildEntity>(cfg =>
            {
                cfg.Property(y => y.Id).ValueGeneratedNever();
                cfg.Property(y => y.Name);
            });
        }
    }

    public class Issue547
    {
        [Theory]
        [InlineData(DbServer.SqlServer)]
        public async Task Test(DbServer dbServer)
        {
            ContextUtil.DbServer = dbServer;

            using var db = new Issue547DbContext(ContextUtil.GetOptions<Issue547DbContext>(databaseName: $"{nameof(EFCoreBulkTest)}_Issue547"));
            await db.Database.EnsureCreatedAsync();


            var tranches = new List<RootEntity>
            {
                new RootEntity {
                    Id = 1,
                    Owned = new OwnedType
                    {
                        Id = 1337,
                        Field1 = "F1",
                        ChildId = 1,
                        Child = new ChildEntity { Id = 1, Name = "F1C1" }
                    }
                },
                new RootEntity {
                    Id = 2,
                    Owned = new OwnedType
                    {
                        Id = 1338,
                        Field1 = "F2",
                        ChildId = 2,
                        Child = new ChildEntity { Id = 2, Name = "F2C2" }
                    }
                }
            };

            await db.BulkInsertOrUpdateAsync(tranches, new BulkConfig
            {
                IncludeGraph = true
            });

            foreach (var a in tranches)
            {
                Assert.True(a.Owned.ChildId.HasValue);
            }
        }
    }
}
```
Nonetheless, I have resolved the (actual) issue in this branch here:
https://github.com/maitlandmarshall/EFCore.BulkExtensions/tree/issue-547

Can you please give it a try and let me know how you go? If it doesn't work, please prepare a unit test which demonstrates the issue.

**Comment by AhmedNetDev** [comment]
Hi @borisdj and thank you again for your time !

Good news, it's working now with OwnsOne & HasOne !

In your unit test, you got the error because you are specifiying the ChildId value in the model, this value will be populated by EF core from the Id of the ChildEntity. So if you delete in the entry model the ChildId value and use the configuration below it will work 👍 

`            modelBuilder.Entity<RootEntity>(cfg =>
            {
                cfg.Property(y => y.Id).ValueGeneratedNever();

                cfg
                  .OwnsOne(t => t.Owned, own =>
                  {
                      own.HasOne(d => d.Child).WithMany().HasForeignKey(d => d.ChildId).IsRequired(false);
                      own.Property(d => d.Field1).HasMaxLength(50);
                  });
            });

            modelBuilder.Entity<ChildEntity>(cfg =>
            {
                cfg.Property(y => y.Id).ValueGeneratedNever();
                cfg.Property(y => y.Name);
            });

            // Entry model
            var tranches = new List<RootEntity>
            {
                new RootEntity {
                    Id = 1,
                    Owned = new OwnedType
                    {
                        Id = 1337,
                        Field1 = "F1",
                        Child = new ChildEntity { Id = 1, Name = "F1C1" }
                    }
                },
                new RootEntity {
                    Id = 2,
                    Owned = new OwnedType
                    {
                        Id = 1338,
                        Field1 = "F2",
                        Child = new ChildEntity { Id = 2, Name = "F2C2" }
                    }
                }
            };
`
 In my case, i understand why i m having the  "random" behavior : I was following DDD by specifying the OwnedTypes as ValueObject with a private constructors. if i remove that all works correctly ! the constructor have to be public.

**So i can confirm that you library is working fine with my specific OwnsOne structure !**

**Please tell me if you confirm also from your side ?**

**Comment by borisdj** [comment]
@AhmedNetDev I believe you are addressing @maitlandmarshall, not me.

**Comment by AhmedNetDev** [comment]
Oups, sorry !

@borisdj :

Good news, it's working now with OwnsOne & HasOne !

In your unit test, you got the error because you are specifiying the ChildId value in the model, this value will be populated by EF core from the Id of the ChildEntity. So if you delete in the entry model the ChildId value and use the configuration below it will work 👍

`
            //EF core configuration
            modelBuilder.Entity<RootEntity>(cfg =>
            {
            cfg.Property(y => y.Id).ValueGeneratedNever();

            cfg
              .OwnsOne(t => t.Owned, own =>
              {
                  own.HasOne(d => d.Child).WithMany().HasForeignKey(d => d.ChildId).IsRequired(false);
                  own.Property(d => d.Field1).HasMaxLength(50);
              });
        });

        modelBuilder.Entity<ChildEntity>(cfg =>
        {
            cfg.Property(y => y.Id).ValueGeneratedNever();
            cfg.Property(y => y.Name);
        });

        // Entry model
        var tranches = new List<RootEntity>
        {
            new RootEntity {
                Id = 1,
                Owned = new OwnedType
                {
                    Id = 1337,
                    Field1 = "F1",
                    Child = new ChildEntity { Id = 1, Name = "F1C1" }
                }
            },
            new RootEntity {
                Id = 2,
                Owned = new OwnedType
                {
                    Id = 1338,
                    Field1 = "F2",
                    Child = new ChildEntity { Id = 2, Name = "F2C2" }
                }
            }
        };
`
In my case, i understand why i m having the "random" behavior : I was following DDD by specifying the OwnedTypes as ValueObject with a private constructors. if i remove that all works correctly ! the constructor have to be public.

So i can confirm that your library is working fine with my specific OwnsOne structure !

Please tell me if you confirm also from your side ?

**Comment by AhmedNetDev** [comment]
> @AhmedNetDev I am unable to replicate the exact issue you are describing. Are you sure an exception is not being thrown?
> 
> When I use OwnsOne it completely fails with an exception.
> `'The MERGE statement conflicted with the FOREIGN KEY constraint "FK_RootEntities_ChildEntity_Owned_ChildId". The conflict occurred in database "EFCoreBulkTest_Issue547", table "dbo.ChildEntity", column 'Id'.`
> 
> This is because the OwnedType confuses the topological sort and inserts the entities in the wrong order.
> 
> Everything imports successfully when I use HasOne instead of OwnsOne.
> 
> ```cs
> namespace EFCore.BulkExtensions.Tests.IncludeGraph
> {
>     public class RootEntity
>     {
>         public int Id { get; set; }
>         public OwnedType Owned { get; set; }
>     }
> 
>     public class OwnedType
>     {
>         public int Id { get; set; }
>         public int? ChildId { get; set; }
>         public ChildEntity Child { get; set; }
>         public string Field1 { get; set; }
>     }
> 
>     public class ChildEntity
>     {
>         public int Id { get; set; }
>         public string Name { get; set; }
>     }
> 
>     public class Issue547DbContext : DbContext
>     {
>         public Issue547DbContext([NotNullAttribute] DbContextOptions options) : base(options)
>         {
>         }
> 
>         public DbSet<RootEntity> RootEntities { get; set; }
> 
>         protected override void OnModelCreating(ModelBuilder modelBuilder)
>         {
>             base.OnModelCreating(modelBuilder);
> 
>             modelBuilder.Entity<RootEntity>(cfg =>
>             {
>                 cfg.Property(y => y.Id).ValueGeneratedNever();
>                 cfg.HasOne(y => y.Owned);
>             });
> 
>             modelBuilder.Entity<OwnedType>(own =>
>             {
>                 own.Property(y => y.Id).ValueGeneratedNever();
>                 own.HasOne(y => y.Child).WithMany().HasForeignKey(y => y.ChildId).IsRequired(false);
>                 own.Property(y => y.Field1).HasMaxLength(50);
>             });
> 
>             modelBuilder.Entity<ChildEntity>(cfg =>
>             {
>                 cfg.Property(y => y.Id).ValueGeneratedNever();
>                 cfg.Property(y => y.Name);
>             });
>         }
>     }
> 
>     public class Issue547
>     {
>         [Theory]
>         [InlineData(DbServer.SqlServer)]
>         public async Task Test(DbServer dbServer)
>         {
>             ContextUtil.DbServer = dbServer;
> 
>             using var db = new Issue547DbContext(ContextUtil.GetOptions<Issue547DbContext>(databaseName: $"{nameof(EFCoreBulkTest)}_Issue547"));
>             await db.Database.EnsureCreatedAsync();
> 
> 
>             var tranches = new List<RootEntity>
>             {
>                 new RootEntity {
>                     Id = 1,
>                     Owned = new OwnedType
>                     {
>                         Id = 1337,
>                         Field1 = "F1",
>                         ChildId = 1,
>                         Child = new ChildEntity { Id = 1, Name = "F1C1" }
>                     }
>                 },
>                 new RootEntity {
>                     Id = 2,
>                     Owned = new OwnedType
>                     {
>                         Id = 1338,
>                         Field1 = "F2",
>                         ChildId = 2,
>                         Child = new ChildEntity { Id = 2, Name = "F2C2" }
>                     }
>                 }
>             };
> 
>             await db.BulkInsertOrUpdateAsync(tranches, new BulkConfig
>             {
>                 IncludeGraph = true
>             });
> 
>             foreach (var a in tranches)
>             {
>                 Assert.True(a.Owned.ChildId.HasValue);
>             }
>         }
>     }
> }
> ```
> 
> Nonetheless, I have resolved the (actual) issue in this branch here:
> https://github.com/maitlandmarshall/EFCore.BulkExtensions/tree/issue-547
> 
> Can you please give it a try and let me know how you go? If it doesn't work, please prepare a unit test which demonstrates the issue.

**Comment by AhmedNetDev** [comment]
@borisdj :

Good news, it's working now with OwnsOne & HasOne !

In your unit test, you got the error because you are specifiying the ChildId value in the model, this value will be populated by EF core from the Id of the ChildEntity. So if you delete in the entry model the ChildId value and use the configuration below it will work 👍

`
        //EF core configuration
        modelBuilder.Entity<RootEntity>(cfg =>
        {
        cfg.Property(y => y.Id).ValueGeneratedNever();

        cfg
          .OwnsOne(t => t.Owned, own =>
          {
              own.HasOne(d => d.Child).WithMany().HasForeignKey(d => d.ChildId).IsRequired(false);
              own.Property(d => d.Field1).HasMaxLength(50);
          });
    });

    modelBuilder.Entity<ChildEntity>(cfg =>
    {
        cfg.Property(y => y.Id).ValueGeneratedNever();
        cfg.Property(y => y.Name);
    });

    // Entry model
    var tranches = new List<RootEntity>
    {
        new RootEntity {
            Id = 1,
            Owned = new OwnedType
            {
                Id = 1337,
                Field1 = "F1",
                Child = new ChildEntity { Id = 1, Name = "F1C1" }
            }
        },
        new RootEntity {
            Id = 2,
            Owned = new OwnedType
            {
                Id = 1338,
                Field1 = "F2",
                Child = new ChildEntity { Id = 2, Name = "F2C2" }
            }
        }
    };
`
In my case, i understand why i m having the "random" behavior : I was following DDD by specifying the OwnedTypes as ValueObject with a private constructors. if i remove that all works correctly ! the constructor have to be public.

**So i can confirm that your library is working fine with my specific OwnsOne structure !

Please tell me if you confirm also from your side ?**

**Comment by borisdj** [comment]
PR merged and v5.2.1 published.
